hazelcast/src/main/java/com/hazelcast/json/internal/JsonSchemaHelper.java

```diff
@@ -0,0 +1,229 @@
+     * @throws IOException
+     */
+    @SuppressWarnings("checkstyle:npathcomplexity")
+    public static JsonPattern createPattern(NavigableJsonInputAdapter input, JsonSchemaNode schemaNode, JsonPathCursor path) {
+        if (schemaNode.isTerminal()) {
+            return null;
+        }
+        JsonPattern pattern = new JsonPattern();
+        while (path.getNext() != null) {
+            if (schemaNode.isTerminal()) {
+                return null;
+            }
+            int suggestedIndexInPath;
+            int numberOfChildren = ((JsonSchemaStructNode) schemaNode).getChildCount();
+            if (path.isArray()) {
+                if (path.isAny()) {
+                    pattern.addAny();
+                    return pattern;
+                }
+                suggestedIndexInPath = path.getArrayIndex();
+            } else {
+                for (suggestedIndexInPath = 0; suggestedIndexInPath < numberOfChildren; suggestedIndexInPath++) {
+                    JsonSchemaNameValue nameValue = ((JsonSchemaStructNode) schemaNode)
+                            .getChild(suggestedIndexInPath);
+                    int nameAddress = nameValue.getNameStart();
+                    if (nameAddress < 0) {
+                        return null;
+                    }
+                    input.position(nameAddress);
+                    if (input.isAttributeName(path)) {
+                        break;
+                    }
+                }
+            }
+            if (isValidIndex(suggestedIndexInPath, (JsonSchemaStructNode) schemaNode, path.isArray())) {
+                schemaNode = ((JsonSchemaStructNode) schemaNode).getChild(suggestedIndexInPath).getValue();
+                pattern.add(suggestedIndexInPath);
+            } else {
+                return null;
+            }
+        }
+        if (schemaNode.isTerminal()) {
+            return pattern;
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Extract the JsonValue that is stored in attributePath in input.
+     * This method validates expectedPattern using attributePath while
+     * extraction. If expectedPattern points to invalid or wrong attributes
+     * at any point, this method returns {@code null}
+     *
+     * NOTE: this method cannot handle patterns with "any" in it.
+     *
+     * @param input             a byte array containing the target object
+     * @param schemaNode        valid schema description to the target
+     *                          object. The behavior is undefined if
+     *                          description does not match the actual
+     *                          object
+     * @param expectedPattern   this cannot contain "any"
+     * @param attributePath     this cannot contain "any"
+     * @return                  JsonValue extracted or null
+     */
+    @SuppressWarnings("checkstyle:npathcomplexity")
+    public static JsonValue findValueWithPattern(NavigableJsonInputAdapter input,
```

Comment:
Here, we wouldn't even need the `attributePath` if the `expectedPattern` matched the path, right? But they can differ?

Reply:
Yes and yes